lib/SILOptimizer/Transforms/TempRValueElimination.cpp

```diff
@@ -441,7 +467,18 @@ bool TempRValueOptPass::tryOptimizeCopyIntoTemp(CopyAddrInst *copyInst) {
   }
 
   // Check if the source is modified within the lifetime of the temporary.
-  if (!checkNoSourceModification(copyInst, copySrc, loadInsts))
+  SILInstruction *lastLoadInst = checkNoSourceModification(copyInst,
+                                                           copySrc, loadInsts);
+  if (!lastLoadInst)
+    return false;
+
+  bool needToInsertDestroy = copyInst->isTakeOfSrc();
```

Comment:
I would also add that in a comment that we are going to insert a destroy_addr in the appropriate spot after rewriting, but before deleting instructions rather than rewrite instructions to ensure proper destruction.

Reply:
I added a comment